pkg/apiserver/webhooks/pvc-mutate.go

```diff
@@ -0,0 +1,89 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package webhooks
+
+import (
+	"context"
+	"encoding/json"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	v1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	"kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	cdiclient "kubevirt.io/containerized-data-importer/pkg/client/clientset/versioned"
+	"kubevirt.io/containerized-data-importer/pkg/common"
+	dvc "kubevirt.io/containerized-data-importer/pkg/controller/datavolume"
+)
+
+type pvcMutatingWebhook struct {
+	k8sClient kubernetes.Interface
+	cdiClient cdiclient.Interface
+}
+
+func (wh *pvcMutatingWebhook) Admit(ar admissionv1.AdmissionReview) *admissionv1.AdmissionResponse {
+	if ar.Request.Operation != admissionv1.Create {
+		return allowedAdmissionResponse()
+	}
+
+	if err := validatePvcResource(ar); err != nil {
+		return toAdmissionResponseError(err)
+	}
+
+	pvc := &v1.PersistentVolumeClaim{}
+	if err := json.Unmarshal(ar.Request.Object.Raw, &pvc); err != nil {
+		return toAdmissionResponseError(err)
+	}
+
+	// Note the webhook LabelSelector should not pass us such pvcs
+	if pvc.Labels[common.PvcUseStorageProfileLabel] != "true" {
+		return allowedAdmissionResponse()
+	}
+
+	pvcCpy := pvc.DeepCopy()
+	if err := dvc.RenderPvc(context.TODO(), wh, pvcCpy); err != nil {
+		return toAdmissionResponseError(err)
+
+	}
+
+	return toPatchResponse(pvc, pvcCpy)
+}
+
+func (wh *pvcMutatingWebhook) GetCDIConfig(ctx context.Context, name string) (*v1beta1.CDIConfig, error) {
+	return wh.cdiClient.CdiV1beta1().CDIConfigs().Get(ctx, common.ConfigName, metav1.GetOptions{})
+}
+
+func (wh *pvcMutatingWebhook) GetStorageClass(ctx context.Context, name string) (*storagev1.StorageClass, error) {
+	return wh.k8sClient.StorageV1().StorageClasses().Get(ctx, name, metav1.GetOptions{})
+}
+
+func (wh *pvcMutatingWebhook) GetStorageProfile(ctx context.Context, name string) (*v1beta1.StorageProfile, error) {
+	return wh.cdiClient.CdiV1beta1().StorageProfiles().Get(ctx, name, metav1.GetOptions{})
+}
+
+func (wh *pvcMutatingWebhook) ListStorageClasses(ctx context.Context) (*storagev1.StorageClassList, error) {
+	return wh.k8sClient.StorageV1().StorageClasses().List(ctx, metav1.ListOptions{})
+}
+
+func (wh *pvcMutatingWebhook) ListPersistentVolumes(ctx context.Context) (*v1.PersistentVolumeList, error) {
+	return wh.k8sClient.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
```

Comment:
Note these are not cached client so those list calls will be costly at scale ("real" clusters)

Reply:
Do we have a way to use the cached client from the webhook?